NetKet/Graph/custom_graph.cc

```diff
@@ -0,0 +1,163 @@
+// Copyright 2018 The Simons Foundation, Inc. - All Rights Reserved.
+// Copyright 2018 Tom Westerhout
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "Graph/custom_graph.hpp"
+
+#include <array>
+#include <algorithm>
+#include <numeric>
+#include <sstream>
+#include <vector>
+#include "Utils/messages.hpp"
+#include "config.hpp"
+
+namespace netket {
+
+namespace {
+auto DefaultAutomorphisms(std::size_t const size)
+    -> std::vector<std::vector<int>> {
+  std::vector<int> identity(size);
+  std::iota(std::begin(identity), std::end(identity), 0);
+  return std::vector<std::vector<int>>{std::move(identity)};
+}
+}  // namespace
+
+NETKET_EXPORT
+CustomGraph::CustomGraph(int const size, bool const is_bipartite)
+    : adjlist_{},
+      eclist_{},
+      nsites_{size},
+      automorphisms_{},
+      isbipartite_{},
+      is_connected_{} {
+  Init(nonstd::nullopt /*adjacency list*/, nonstd::nullopt /*automorphisms*/,
+       is_bipartite, nullptr);
+}
+
+NETKET_EXPORT
+CustomGraph::CustomGraph(int const size,
+                         std::vector<std::vector<int>> automorphisms,
+                         bool const is_bipartite)
+    : adjlist_{},
+      eclist_{},
+      nsites_{size},
+      automorphisms_{},
+      isbipartite_{},
+      is_connected_{} {
+  Init(nonstd::nullopt /*adjacency list*/, std::move(automorphisms),
+       is_bipartite, nullptr);
+}
+
+NETKET_EXPORT
+CustomGraph::CustomGraph(std::vector<std::vector<int>> adjlist,
+                         bool const is_bipartite,
+                         std::vector<std::vector<int>> const *colorlist)
+    : adjlist_{},
+      eclist_{},
+      nsites_{},
+      automorphisms_{},
+      isbipartite_{},
+      is_connected_{} {
+  Init(std::move(adjlist), nonstd::nullopt /*automorphisms*/, is_bipartite,
+       colorlist);
+}
+
+NETKET_EXPORT
+CustomGraph::CustomGraph(std::vector<std::vector<int>> adjlist,
+                         std::vector<std::vector<int>> automorphisms,
+                         bool const is_bipartite,
+                         std::vector<std::vector<int>> const *colorlist)
```

Comment:
No, @gcarleo I actually like the idea! No need for the `size != 0` check I think... We can even do the same thing with `automorphisms`, because we use identity by default, i.e. `automorphism.size() > 0`. This should reduce the number of constructors to 1!